presto-main/src/main/java/io/prestosql/operator/scalar/time/TimeOperators.java

```diff
@@ -0,0 +1,298 @@
+            long picos = parseTime(value.toStringUtf8());
+            return round(picos, (int) (MAX_PRECISION - precision)) % PICOSECONDS_PER_DAY;
+        }
+        catch (IllegalArgumentException e) {
+            throw new PrestoException(INVALID_CAST_ARGUMENT, "Value cannot be cast to time: " + value.toStringUtf8(), e);
+        }
+    }
+
+    @ScalarOperator(HASH_CODE)
+    @LiteralParameters("p")
+    @SqlType(StandardTypes.BIGINT)
+    public static long hashCode(@SqlType("time(p)") long value)
+    {
+        return AbstractLongType.hash(value);
+    }
+
+    @ScalarOperator(XX_HASH_64)
+    @LiteralParameters("p")
+    @SqlType(StandardTypes.BIGINT)
+    public static long xxHash64(@SqlType("time(p)") long value)
+    {
+        return XxHash64.hash(value);
+    }
+
+    @ScalarOperator(IS_DISTINCT_FROM)
+    public static final class TimeDistinctFromOperator
+    {
+        @LiteralParameters("p")
+        @SqlType(StandardTypes.BOOLEAN)
+        public static boolean isDistinctFrom(
+                @SqlType("time(p)") long left,
+                @IsNull boolean leftNull,
+                @SqlType("time(p)") long right,
+                @IsNull boolean rightNull)
+        {
+            if (leftNull != rightNull) {
+                return true;
+            }
+            if (leftNull) {
+                return false;
+            }
+            return notEqual(left, right);
+        }
+
+        @LiteralParameters("p")
+        @SqlType(StandardTypes.BOOLEAN)
+        public static boolean isDistinctFrom(
+                @BlockPosition @SqlType(value = "time(p)", nativeContainerType = long.class) Block left,
+                @BlockIndex int leftPosition,
+                @BlockPosition @SqlType(value = "time(p)", nativeContainerType = long.class) Block right,
+                @BlockIndex int rightPosition)
+        {
+            if (left.isNull(leftPosition) != right.isNull(rightPosition)) {
+                return true;
+            }
+            if (left.isNull(leftPosition)) {
+                return false;
+            }
+            return notEqual(TIME.getLong(left, leftPosition), TIME.getLong(right, rightPosition));
+        }
+    }
+
+    @ScalarOperator(INDETERMINATE)
+    @LiteralParameters("p")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean indeterminate(@SqlType("time(p)") long value, @IsNull boolean isNull)
+    {
+        return isNull;
+    }
+
+    @ScalarOperator(CAST)
+    @LiteralParameters({"sourcePrecision", "targetPrecision"})
+    @SqlType("time(targetPrecision)")
+    public static long castToTime(
+            @LiteralParameter("sourcePrecision") long sourcePrecision,
+            @LiteralParameter("targetPrecision") long targetPrecision,
+            @SqlType("time(sourcePrecision)") long time)
+    {
+        if (sourcePrecision <= targetPrecision) {
+            return time;
+        }
+
+        return round(time, (int) (MAX_PRECISION - targetPrecision)) % PICOSECONDS_PER_DAY;
```

Comment:
`% PICOSECONDS_PER_DAY` seems redundant

add a comment if it's not

Reply:
It's needed because `round` can round up. So a time like `23:59:59.9` will round to `24:00:00`, which needs to turn into `00:00:00`

Reply:
Co `cast(TIME '23:59:59.9' AS time(0))` returns `00:00:00` rather than closes representable value, `23:59:59`?

Reply:
Yes, all operations on time are done modulo 24h (as required by the spec). Since we use rounding semantics, that's the reasonable thing to do.

Reply:
I bet this will bite us. Maybe just because this is behavior that I could hardly foresee as possible.

Reply:
That's just how the type works in the spec. It says: 

> b) If SD is TIME WITHOUT TIME ZONE, then TV is SV, with implementation-defined rounding or
truncation if necessary.

Also, elsewhere:

> f the data type of the <datetime term> or <datetime value expression> is time with or without time zone, then arithmetic on the HOUR <primary datetime field> is undertaken modulo 24. 

Since we chose rounding, then this behavior follows from it. This is consistent with this, too:

```
presto:default> select cast(timestamp '2020-01-01 23:59:59.999999' as timestamp(0));
        _col0
---------------------
 2020-01-02 00:00:00
(1 row)
```

It would be very surprising if these expressions produced different results:

```
CAST(CAST(timestamp '2020-01-01 23:59:59.999999' as timestamp(0)) AS time(0))
CAST(CAST(timestamp '2020-01-01 23:59:59.999999' AS time(6)) AS time(0))
CAST(timestamp '2020-01-01 23:59:59.999999' AS time(0))
```

Reply:
Thanks @martint for taking time to explain this.
It's seem consequential...
Do you think we could document this somewhere?
Docs or internally? I bet I or someone else will "rediscover" this some day,
and would be nice to be able to refer to something.

Reply:
FWIW

### PostgreSQL 12.3
```
test=# SELECT CAST(TIME '23:59:59.999' AS time(1));
   time
----------
 24:00:00
(1 row)
```

### Oracle 11g r2 XE
```
SQL> SELECT CAST(TIME '23:59:59.999' AS time(1)) FROM dual;

CAST(TIME'23:59:59.999'ASTIME(1))
---------------------------------------------------------------------------
12.00.00.0 AM
```

12 AM is 00:00:00

```
SQL> SELECT to_char(CAST(TIME '23:59:59.999' AS time(1)), 'hh24:mi:ss') FROM dual;

TO_CHAR(
--------
00:00:00
```

### SQL Server 2017

```
1> SELECT CAST('23:59:59.999' AS time);
2> GO

----------------------
      23:59:59.9990000
```

```
1> SELECT CAST(CAST('23:59:59.999' AS time) AS time(1));
2> GO

----------------------
            23:59:59.9
```